{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## DAISY- the DAta-mIning SYnthetic-lethality-identification pipeline\n",
    "```\n",
    "Title:   Data Mining Synthetic Lethality Identification Pipeline (DAISY)\n",
    "Author:  Bahar Tercan\n",
    "Created: 02-07-2022\n",
    "Purpose: Retrive Synthetic Lethal Partners of The Genes in the Given List Using DAISY Algorithm \n",
    "Notes: Runs in MyBinder \n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
   "If any piece of this analysis is used in a publication please cite the following:\n",
    "\n",
    "For DAISY algorithm: \n",
    "\n",
    "Jerby-Arnon, L., Pfetzer, N., Waldman, Y. Y., McGarry, L., James, D., Shanks, E., ... & Gottlieb, E. (2014). Predicting cancer-specific vulnerability via data-driven detection of synthetic lethality. Cell, 158(5), 1199-1209.\n",
    "\n",
    "For CCLE Omics data:\n",
    "\n",
    "Ghandi, M., Huang, F.W., Jané-Valbuena, J. et al. Next-generation characterization of the Cancer Cell Line Encyclopedia. Nature 569, 503–508 (2019). https://doi.org/10.1038/s41586-019-1186-3\n",
    "\n",
    "For CRISPR Data: \n",
    "\n",
    "Robin M. Meyers, Jordan G. Bryan, James M. McFarland, Barbara A. Weir, ... David E. Root, William C. Hahn, Aviad Tsherniak. Computational correction of copy number effect improves specificity of CRISPR-Cas9 essentiality screens in cancer cells. Nature Genetics 2017 October 49:1779–1784. doi:10.1038/ng.3984\n",
    "\n",
    "Dempster, J. M., Rossen, J., Kazachkova, M., Pan, J., Kugener, G., Root, D. E., & Tsherniak, A. (2019). Extracting Biological Insights from the Project Achilles Genome-Scale CRISPR Screens in Cancer Cell Lines. BioRxiv, 720243.\n",
    "\n",
    "For shRNA Data:\n",
    "\n",
    "James M. McFarland, Zandra V. Ho, Guillaume Kugener, Joshua M. Dempster, Phillip G. Montgomery, Jordan G. Bryan, John M. Krill-Burger, Thomas M. Green, Francisca Vazquez, Jesse S. Boehm, Todd R. Golub, William C. Hahn, David E. Root, Aviad Tsherniak. (2018). Improved estimation of cancer dependencies from large-scale RNAi screens using model-based normalization and data integration. Nature Communications 9, 1. https://doi.org/10.1038/s41467-018-06916-5\n",
    "\n",
    "For ISB-CGC:\n",
     "\n",
    "Reynolds, S. M., Miller, M., Lee, P., Leinonen, K., Paquette, S. M., Rodebaugh, Z., ... & Shmulevich, I. (2017). The ISB Cancer Genomics Cloud: a flexible cloud-based platform for cancer genomics research. Cancer research, 77(21), e7-e10.\n",
    "\n",
    "For Pancancer Atlas Data:\n",
     "\n",
    "Hutter, C., and Zenklusen, J.C. (2018). The Cancer Genome Atlas: Creating Lasting Value beyond Its Data. Cell 173, 283–285.\n",
    "\n",
    "This notebook is a reimplementation of DAISY Synthetic Lethal Pair Prediction Algorithm\n",
    "\n",
    "It consists 3 modules: \n",
    "\n",
    "1. SL/SDL candidate determination using gene co-expression\n",
    "2. SL/SDL candidate determination using survival of fittest\n",
    "3. SL/SDL candidate determination using CRISPR and shRNA experiments\n",
    "\n",
    "* The results from the three modules were then aggregated into one list of candidate SL/SDL pairs\n",
    "\n",
    "Input Parameters\n",
    "* Cancer type \n",
    "* The genes whose SL/SDL partners are seeked\n",
    "\n",
    "Input Data (available in bigquery tables)\n",
    "* Gene expression data \n",
    "* Gene mutation data\n",
    "* Copy number variation data\n",
    "* Gene effect data (CRISPR)\n",
    "* Gene dependency scores data (shRNA)\n",
    "\n",
    "Output\n",
    "* List of candidate SL/SDL pairs\n",
    "\n",
    "Please contact Bahar Tercan btercan@systemsbiology.org for your questions and detailed information. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Please don't run it if you are running the notebook in MyBinder\n",
    "# This code block installs the dependencies, please uncomment the code lines and run it only once, the first time you run this notebook on your computer\n",
    "\n",
    "#!pip3 install google.cloud\n",
    "#!pip3 install importlib\n",
    "#!pip3 install pandas\n",
    "#!pip3 install ipywidgets\n",
    "#!pip3 install numpy\n",
    "#!pip3 install statsmodels\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1. Import python libraries\n",
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.path.append('../Scripts/') # to be able to use the .py files in ../Scripts folder\n",
    "from google.cloud import bigquery\n",
    "import importlib\n",
    "import pandas as pd\n",
    "import DAISY_operations\n",
    "importlib.reload(DAISY_operations)\n",
    "from DAISY_operations import *\n",
    "import ipywidgets as widgets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "if not sys.warnoptions:\n",
    "    import warnings\n",
    "    warnings.simplefilter(\"ignore\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Google Authentication\n",
    "Running the BigQuery cells in this notebook requires a Google Cloud Project, instructions for creating a project can be found in the [Google Documentation](https://cloud.google.com/resource-manager/docs/creating-managing-projects#console). The instance needs to be authorized to bill the project for queries.\n", 
    "For more information on getting started in the cloud see ['Quick Start Guide to ISB-CGC'](https://isb-cancer-genomics-cloud.readthedocs.io/en/latest/sections/HowToGetStartedonISB-CGC.html) and alternative authentication methods can be found in the [Google Documentation](https://googleapis.dev/python/google-api-core/latest/auth.html)."
   ] 
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Please don't run this code block if you are running the notebook in MyBinder\n",
    "#If you are not running the notebook in MyBinder, please make sure that you have installed Cloud SDK.\n",
    "#See support from https://cloud.google.com/sdk/docs/install\n",
    "\n",
    "!gcloud auth application-default login\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2. Sign in Google Bigquery with the project id\n",
    "\n",
    "Bigquery connection\n",
    "\n"   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#Please don't run this code block if you are running the notebook in MyBinder\n",
    "#Please replace 'syntheticlethality' with your project id\n",
    "project_id='syntheticlethality'\n",
    "client = bigquery.Client(project_id)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 4. Prediction of synthetic lethal partners using different modules on DAISY\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There are three modules for synthetic lethal/synthetic dosage lethal pair inferences on DAISY :\n",
    "1. Pairwise gene coexpression, \n",
    "\n",
    "2. Genomic survival of the fittest. \n",
    "\n",
    "3. shRNA or CRISPR based functional examination.\n",
    "You can get more information in the original paper : https://www.sciencedirect.com/science/article/pii/S0092867414009775.\n",
    "\n",
    "In pairwise gene coexpression module and genomic survial of the fittest module, we will use PancancerAtlas and CCLE data.<br>\n",
    "In functional examination module, we will use CRISPR and shRNA data together with CCLE data. <br>\n",
    "\n",
    "Python codes required are  in the ../Scripts/ folder and they are imported at the beginning. \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 4.0. Default parameters for DAISY, you can edit them\n",
    "\n",
    "For SDL prediction, please replace 'SL' with 'SDL' and 'Inactive' with 'Overactive' in the following code lines\n",
    "\n",
    "For SOF and FuncExamination Procedures, input_mutations is an optional parameter, if you don't want to use,  you can skip "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "input_mutations = ['Nonsense_Mutation', 'Frame_Shift_Ins', 'Frame_Shift_Del'] \n",
    "# DAISY default parameters for SL prediction\n",
    "percentile_threshold = 10 # used for deciding whether the gene is inactive\n",
    "cn_threshold = -0.3 # used for deciding whether the gene is inactive\n",
    "cor_threshold = 0.5 # used for inferring whether two genes are in SL relationship\n",
    "p_threshold = 0.05\n",
    "pval_correction = 'Bonferroni'\n",
    "fdr_level='gene_level' #it can be gene_level or analysis_level\n",
    "\n",
    "# for SDL prediction DAISY parameters are \n",
    "#percentile_threshold = 90\n",
    "#cn_threshold = 0.3 \n",
    "#cor_threshold = 0.5\n",
    "#p_threshold = 0.05\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The tumor types are the TCGA cancer types, the cancer types that have corresponding Celllines are listed in the combobox. Please click on the tissue(s) you want to do the analyses on."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "14995354d72e45e88a4cb9e21e062f3c",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "SelectMultiple(description='Tumor type', options=('pancancer', 'ALL', 'LAML', 'CML', 'NSCLC', 'SCLC', 'SKCM', …"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "TCGA_list=GetTCGASubtypes(client)\n",
    "TCGA_list = [i for i in TCGA_list if i]\n",
    "\n",
    "tumor_type = widgets.SelectMultiple(\n",
    "    options=['pancancer'] + TCGA_list  ,\n",
    "    value=[],\n",
    "    description='Tumor type',\n",
    "    disabled=False\n",
    ")\n",
    "display(tumor_type)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The list of genes for which we would like to find SL/SDL partners."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_list=[\"BRCA1\", \"BRCA2\", \"ARID1A\"] # any number of genes in list format"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 4.1. Pairwise gene coexpression module\n",
    "\n",
    "In the pairwise co-expression module, DAISY makes inferences based on the assumption that synthetic lethal gene pairs play a role in related biological processes and are co-expressed. Gene expression was measured for patient-derived data from TCGA and CCLE cancer cell lines. Pairwise co-expression is estimated from the Spearman correlation which we calculated between each gene of interest (each item in the query gene list) and all other genes. Candidate synthetic lethal gene pairs are those with correlation coefficient greater than 0.5 and whose Bonferroni-corrected P value was smaller than 0.05 by default, these parameters can be set below. \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "4.1.1. Pairwise gene coexpression module on PancancerAtlas."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>Inactive</th>\n",
       "      <th>InactiveDB</th>\n",
       "      <th>SL_Candidate</th>\n",
       "      <th>#Samples</th>\n",
       "      <th>Correlation</th>\n",
       "      <th>PValue</th>\n",
       "      <th>FDR</th>\n",
       "      <th>Tissue</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Inactive</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">ARID1A</th>\n",
       "      <th>0</th>\n",
       "      <td>ARID1A</td>\n",
       "      <td>ARID1A</td>\n",
       "      <td>SPEN</td>\n",
       "      <td>9953</td>\n",
       "      <td>0.726948</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>ARID1A</td>\n",
       "      <td>ARID1A</td>\n",
       "      <td>ZNF407</td>\n",
       "      <td>9953</td>\n",
       "      <td>0.525198</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td>ARID1A</td>\n",
       "      <td>ARID1A</td>\n",
       "      <td>BOD1L</td>\n",
       "      <td>9953</td>\n",
       "      <td>0.525071</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103</th>\n",
       "      <td>ARID1A</td>\n",
       "      <td>ARID1A</td>\n",
       "      <td>C2CD3</td>\n",
       "      <td>9953</td>\n",
       "      <td>0.524484</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104</th>\n",
       "      <td>ARID1A</td>\n",
       "      <td>ARID1A</td>\n",
       "      <td>ZNF644</td>\n",
       "      <td>9953</td>\n",
       "      <td>0.524323</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">BRCA2</th>\n",
       "      <th>41196</th>\n",
       "      <td>BRCA2</td>\n",
       "      <td>BRCA2</td>\n",
       "      <td>MLF1IP</td>\n",
       "      <td>9953</td>\n",
       "      <td>0.649312</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41197</th>\n",
       "      <td>BRCA2</td>\n",
       "      <td>BRCA2</td>\n",
       "      <td>DHX15</td>\n",
       "      <td>9953</td>\n",
       "      <td>0.649128</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41198</th>\n",
       "      <td>BRCA2</td>\n",
       "      <td>BRCA2</td>\n",
       "      <td>RAD54B</td>\n",
       "      <td>9953</td>\n",
       "      <td>0.649032</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41218</th>\n",
       "      <td>BRCA2</td>\n",
       "      <td>BRCA2</td>\n",
       "      <td>CCNB1</td>\n",
       "      <td>9953</td>\n",
       "      <td>0.638053</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41583</th>\n",
       "      <td>BRCA2</td>\n",
       "      <td>BRCA2</td>\n",
       "      <td>DNMT3B</td>\n",
       "      <td>9953</td>\n",
       "      <td>0.500582</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1166 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               Inactive InactiveDB SL_Candidate  #Samples  Correlation  \\\n",
       "Inactive                                                                 \n",
       "ARID1A   0       ARID1A     ARID1A         SPEN      9953     0.726948   \n",
       "         101     ARID1A     ARID1A       ZNF407      9953     0.525198   \n",
       "         102     ARID1A     ARID1A        BOD1L      9953     0.525071   \n",
       "         103     ARID1A     ARID1A        C2CD3      9953     0.524484   \n",
       "         104     ARID1A     ARID1A       ZNF644      9953     0.524323   \n",
       "...                 ...        ...          ...       ...          ...   \n",
       "BRCA2    41196    BRCA2      BRCA2       MLF1IP      9953     0.649312   \n",
       "         41197    BRCA2      BRCA2        DHX15      9953     0.649128   \n",
       "         41198    BRCA2      BRCA2       RAD54B      9953     0.649032   \n",
       "         41218    BRCA2      BRCA2        CCNB1      9953     0.638053   \n",
       "         41583    BRCA2      BRCA2       DNMT3B      9953     0.500582   \n",
       "\n",
       "                PValue  FDR         Tissue  \n",
       "Inactive                                    \n",
       "ARID1A   0         0.0  0.0  ['pancancer']  \n",
       "         101       0.0  0.0  ['pancancer']  \n",
       "         102       0.0  0.0  ['pancancer']  \n",
       "         103       0.0  0.0  ['pancancer']  \n",
       "         104       0.0  0.0  ['pancancer']  \n",
       "...                ...  ...            ...  \n",
       "BRCA2    41196     0.0  0.0  ['pancancer']  \n",
       "         41197     0.0  0.0  ['pancancer']  \n",
       "         41198     0.0  0.0  ['pancancer']  \n",
       "         41218     0.0  0.0  ['pancancer']  \n",
       "         41583     0.0  0.0  ['pancancer']  \n",
       "\n",
       "[1166 rows x 8 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "coexp_pancancer = CoexpressionAnalysis(client, 'SL', \"PanCancerAtlas\", gene_list , pval_correction, fdr_level, list(tumor_type.value))\n",
    "try:\n",
    "    coex_pan_intermediate_report=coexp_pancancer.loc[(coexp_pancancer['FDR'] < p_threshold)&(coexp_pancancer['Correlation'] > cor_threshold)]\n",
    "    coexp_pancancer_report=coex_pan_intermediate_report.groupby('Inactive').apply(lambda x: x.sort_values('FDR'))\n",
    "except:\n",
    "    coexp_pancancer_report=pd.DataFrame()\n",
    "    print(\"No results returned.\")\n",
    "    \n",
    "coexp_pancancer_report"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Inactive/Overactive** (for SL and for SDL respectively): The gene name coming from the input list\n",
    "\n",
    "**InactiveDB/OveractiveDB**: The corresponding gene name in the dataset (Pancancer Atlas)\n",
    "\n",
    "**SL_Candidate**: The Candidate SL pair of the gene in the same row \n",
    "\n",
    "**#Samples**: The number of samples that have been used in Correlation Calculation\n",
    "\n",
    "**Correlation**: The Spearman correlation rho on gene expression.\n",
    "\n",
    "**PValue**: The Spearman correlation p value\n",
    "\n",
    "**FDR**\t:Corrected p value \n",
    "\n",
    "**Tissue**: The tissue(s) that the analysis has been performed. \n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "4.1.2. Pairwise gene coexpression module on CCLE data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Inactive</th>\n",
       "      <th>InactiveDB</th>\n",
       "      <th>SL_Candidate</th>\n",
       "      <th>#Samples</th>\n",
       "      <th>Correlation</th>\n",
       "      <th>PValue</th>\n",
       "      <th>FDR</th>\n",
       "      <th>Tissue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ARID1A</td>\n",
       "      <td>ARID1A</td>\n",
       "      <td>HCFC1</td>\n",
       "      <td>1297</td>\n",
       "      <td>0.779066</td>\n",
       "      <td>8.589353e-265</td>\n",
       "      <td>1.642198e-260</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ARID1A</td>\n",
       "      <td>ARID1A</td>\n",
       "      <td>SPEN</td>\n",
       "      <td>1297</td>\n",
       "      <td>0.765432</td>\n",
       "      <td>4.079862e-250</td>\n",
       "      <td>7.800289e-246</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ARID1A</td>\n",
       "      <td>ARID1A</td>\n",
       "      <td>SF1</td>\n",
       "      <td>1297</td>\n",
       "      <td>0.763544</td>\n",
       "      <td>3.668113e-248</td>\n",
       "      <td>7.013066e-244</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ARID1A</td>\n",
       "      <td>ARID1A</td>\n",
       "      <td>UBTF</td>\n",
       "      <td>1297</td>\n",
       "      <td>0.748176</td>\n",
       "      <td>6.490310e-233</td>\n",
       "      <td>1.240882e-228</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ARID1A</td>\n",
       "      <td>ARID1A</td>\n",
       "      <td>KMT2D</td>\n",
       "      <td>1297</td>\n",
       "      <td>0.733084</td>\n",
       "      <td>5.556467e-219</td>\n",
       "      <td>1.062341e-214</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38488</th>\n",
       "      <td>BRCA2</td>\n",
       "      <td>BRCA2</td>\n",
       "      <td>SUV39H2</td>\n",
       "      <td>1297</td>\n",
       "      <td>0.501409</td>\n",
       "      <td>1.681160e-83</td>\n",
       "      <td>3.213873e-79</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38489</th>\n",
       "      <td>BRCA2</td>\n",
       "      <td>BRCA2</td>\n",
       "      <td>RRM1</td>\n",
       "      <td>1297</td>\n",
       "      <td>0.501355</td>\n",
       "      <td>1.760925e-83</td>\n",
       "      <td>3.366359e-79</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38490</th>\n",
       "      <td>BRCA2</td>\n",
       "      <td>BRCA2</td>\n",
       "      <td>SMARCA5</td>\n",
       "      <td>1297</td>\n",
       "      <td>0.501238</td>\n",
       "      <td>1.949504e-83</td>\n",
       "      <td>3.726867e-79</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38491</th>\n",
       "      <td>BRCA2</td>\n",
       "      <td>BRCA2</td>\n",
       "      <td>CEP295</td>\n",
       "      <td>1297</td>\n",
       "      <td>0.501160</td>\n",
       "      <td>2.086518e-83</td>\n",
       "      <td>3.988797e-79</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38492</th>\n",
       "      <td>BRCA2</td>\n",
       "      <td>BRCA2</td>\n",
       "      <td>MTBP</td>\n",
       "      <td>1297</td>\n",
       "      <td>0.500160</td>\n",
       "      <td>4.966547e-83</td>\n",
       "      <td>9.494547e-79</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1737 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Inactive InactiveDB SL_Candidate  #Samples  Correlation         PValue  \\\n",
       "0       ARID1A     ARID1A        HCFC1      1297     0.779066  8.589353e-265   \n",
       "1       ARID1A     ARID1A         SPEN      1297     0.765432  4.079862e-250   \n",
       "2       ARID1A     ARID1A          SF1      1297     0.763544  3.668113e-248   \n",
       "3       ARID1A     ARID1A         UBTF      1297     0.748176  6.490310e-233   \n",
       "4       ARID1A     ARID1A        KMT2D      1297     0.733084  5.556467e-219   \n",
       "...        ...        ...          ...       ...          ...            ...   \n",
       "38488    BRCA2      BRCA2      SUV39H2      1297     0.501409   1.681160e-83   \n",
       "38489    BRCA2      BRCA2         RRM1      1297     0.501355   1.760925e-83   \n",
       "38490    BRCA2      BRCA2      SMARCA5      1297     0.501238   1.949504e-83   \n",
       "38491    BRCA2      BRCA2       CEP295      1297     0.501160   2.086518e-83   \n",
       "38492    BRCA2      BRCA2         MTBP      1297     0.500160   4.966547e-83   \n",
       "\n",
       "                 FDR         Tissue  \n",
       "0      1.642198e-260  ['pancancer']  \n",
       "1      7.800289e-246  ['pancancer']  \n",
       "2      7.013066e-244  ['pancancer']  \n",
       "3      1.240882e-228  ['pancancer']  \n",
       "4      1.062341e-214  ['pancancer']  \n",
       "...              ...            ...  \n",
       "38488   3.213873e-79  ['pancancer']  \n",
       "38489   3.366359e-79  ['pancancer']  \n",
       "38490   3.726867e-79  ['pancancer']  \n",
       "38491   3.988797e-79  ['pancancer']  \n",
       "38492   9.494547e-79  ['pancancer']  \n",
       "\n",
       "[1737 rows x 8 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "coexp_CCLE=CoexpressionAnalysis(client, 'SL', 'CCLE', gene_list, pval_correction, fdr_level, list(tumor_type.value ))\n",
    "try: \n",
    "    coex_ccle_intermediate_report=coexp_CCLE.loc[(coexp_CCLE['FDR'] < p_threshold)&(coexp_CCLE['Correlation'] > cor_threshold)]\n",
    "    coexp_CCLE_report=coex_ccle_intermediate_report.groupby('Inactive').apply(lambda x: x.sort_values('FDR'))\n",
    "except:\n",
    "    coexp_CCLE_report=pd.DataFrame()\n",
    "    print(\"No results returned.\")\n",
    "coexp_CCLE_report\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The same analysis/output with Pancancer Atlas Coexpression analysis except for the dataset used (CCLE). \n",
     "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 4.2. Genomic survival of fittest module\n",
    "The genomic survival of the fittest inference module is based on the copy number alteration of the gene in the search domain, given whether the gene of interest is inactive (overactive) or not.The gene of interest in a sample is considered inactive if its expression is less than 10th percentile across all samples and its copy number alteration is less than -0.3 or if it has a nonsense, frame shift or frame-del mutation. The gene of interest in a sample is considered overactive if it has gene expression exceeding the 90th percentile across all samples and its copy number alteration is greater  than 0.3 (over-activity is used in synthetic dosage lethal pair prediction)\n",
    "The one-sided Wilcoxon rank-sum (Mann-Whitney U) test was applied to the copy number alteration of the candidate synthetic lethal pair of each gene of interest. The higher copy number of the candidate synthetic lethal pair for  the samples whose gene of interest is inactive (overactive) is considered as an indicator of the genes being in a synthetic lethal or synthetic dosage lethal relationship. The SL/SDL pairs whose Bonferroni - corrected p-value is  less than 0.05  were returned. This inference procedure was applied on PanCancer Atlas  and CCLE data separately. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "4.2.1. Genomic survival of fittest module on PancancerAtlas data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>Inactive</th>\n",
       "      <th>InactiveDB</th>\n",
       "      <th>SL_Candidate</th>\n",
       "      <th>#InactiveSamples</th>\n",
       "      <th>#Samples</th>\n",
       "      <th>PValue</th>\n",
       "      <th>FDR</th>\n",
       "      <th>Tissue</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Inactive</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">ARID1A</th>\n",
       "      <th>0</th>\n",
       "      <td>ARID1A</td>\n",
       "      <td>ARID1A</td>\n",
       "      <td>ENTPD7</td>\n",
       "      <td>879</td>\n",
       "      <td>8930</td>\n",
       "      <td>2.053913e-14</td>\n",
       "      <td>5.161072e-10</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ARID1A</td>\n",
       "      <td>ARID1A</td>\n",
       "      <td>SLC25A28</td>\n",
       "      <td>879</td>\n",
       "      <td>8930</td>\n",
       "      <td>2.436940e-14</td>\n",
       "      <td>6.123542e-10</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ARID1A</td>\n",
       "      <td>ARID1A</td>\n",
       "      <td>MYOF</td>\n",
       "      <td>879</td>\n",
       "      <td>8930</td>\n",
       "      <td>2.692291e-14</td>\n",
       "      <td>6.765188e-10</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ARID1A</td>\n",
       "      <td>ARID1A</td>\n",
       "      <td>RNLS</td>\n",
       "      <td>879</td>\n",
       "      <td>8930</td>\n",
       "      <td>2.969847e-14</td>\n",
       "      <td>7.462631e-10</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ARID1A</td>\n",
       "      <td>ARID1A</td>\n",
       "      <td>ANKRD1</td>\n",
       "      <td>879</td>\n",
       "      <td>8930</td>\n",
       "      <td>3.075318e-14</td>\n",
       "      <td>7.727659e-10</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">BRCA1</th>\n",
       "      <th>2813</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>OR2A2</td>\n",
       "      <td>229</td>\n",
       "      <td>8930</td>\n",
       "      <td>1.968141e-06</td>\n",
       "      <td>4.945543e-02</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2815</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>LRRC28</td>\n",
       "      <td>229</td>\n",
       "      <td>8930</td>\n",
       "      <td>1.976030e-06</td>\n",
       "      <td>4.965368e-02</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2816</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>FAM177A1</td>\n",
       "      <td>229</td>\n",
       "      <td>8930</td>\n",
       "      <td>1.976277e-06</td>\n",
       "      <td>4.965989e-02</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2818</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>ZFP82</td>\n",
       "      <td>229</td>\n",
       "      <td>8930</td>\n",
       "      <td>1.983702e-06</td>\n",
       "      <td>4.984646e-02</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2820</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>LENG8</td>\n",
       "      <td>229</td>\n",
       "      <td>8930</td>\n",
       "      <td>1.987797e-06</td>\n",
       "      <td>4.994935e-02</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2821 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              Inactive InactiveDB SL_Candidate  #InactiveSamples  #Samples  \\\n",
       "Inactive                                                                     \n",
       "ARID1A   0      ARID1A     ARID1A       ENTPD7               879      8930   \n",
       "         1      ARID1A     ARID1A     SLC25A28               879      8930   \n",
       "         2      ARID1A     ARID1A         MYOF               879      8930   \n",
       "         3      ARID1A     ARID1A         RNLS               879      8930   \n",
       "         4      ARID1A     ARID1A       ANKRD1               879      8930   \n",
       "...                ...        ...          ...               ...       ...   \n",
       "BRCA1    2813    BRCA1      BRCA1        OR2A2               229      8930   \n",
       "         2815    BRCA1      BRCA1       LRRC28               229      8930   \n",
       "         2816    BRCA1      BRCA1     FAM177A1               229      8930   \n",
       "         2818    BRCA1      BRCA1        ZFP82               229      8930   \n",
       "         2820    BRCA1      BRCA1        LENG8               229      8930   \n",
       "\n",
       "                     PValue           FDR         Tissue  \n",
       "Inactive                                                  \n",
       "ARID1A   0     2.053913e-14  5.161072e-10  ['pancancer']  \n",
       "         1     2.436940e-14  6.123542e-10  ['pancancer']  \n",
       "         2     2.692291e-14  6.765188e-10  ['pancancer']  \n",
       "         3     2.969847e-14  7.462631e-10  ['pancancer']  \n",
       "         4     3.075318e-14  7.727659e-10  ['pancancer']  \n",
       "...                     ...           ...            ...  \n",
       "BRCA1    2813  1.968141e-06  4.945543e-02  ['pancancer']  \n",
       "         2815  1.976030e-06  4.965368e-02  ['pancancer']  \n",
       "         2816  1.976277e-06  4.965989e-02  ['pancancer']  \n",
       "         2818  1.983702e-06  4.984646e-02  ['pancancer']  \n",
       "         2820  1.987797e-06  4.994935e-02  ['pancancer']  \n",
       "\n",
       "[2821 rows x 8 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sof_pancancer = SurvivalOfFittest(client, 'SL', \"PanCancerAtlas\", gene_list, percentile_threshold, cn_threshold, pval_correction, fdr_level, list(tumor_type.value), input_mutations)\n",
    "try:\n",
    "    sof_pancancer_intermediate_report=sof_pancancer.loc[(sof_pancancer['FDR'] < p_threshold),]                \n",
    "    sof_pancancer_report=sof_pancancer_intermediate_report.groupby('Inactive').apply(lambda x: x.sort_values('FDR'))\n",
    "except:\n",
    "    sof_pancancer_report=pd.DataFrame()\n",
    "    print(\"No results returned.\") \n",
    "sof_pancancer_report"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Inactive/Overactive** (for SL and for SDL respectively): The gene name coming from the input list.\n",
    "\n",
    "**InactiveDB/OveractiveDB**: The corresponding gene name in the dataset (Pancancer Atlas).\n",
    "\n",
    "**SL_Candidate**: The Candidate SL pair of the gene in the same row. \n",
    "\n",
    "**#InactiveSamples:** The number of inactive samples.\n",
    "\n",
    "**#Samples:** The total number of samples that the analysis is performed on. \n",
    "\n",
    "**PValue:** The p values for the Wilcoxon unpaired one sided test on Somatic Copy Number Alteration data\n",
    "\n",
    "**FDR:** Corrected p value\n",
    "\n",
    "**Tissue:** The tissue(s) that the analysis has been performed."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    " 4.2.1. Genomic survival of fittest module on CCLE data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Inactive</th>\n",
       "      <th>InactiveDB</th>\n",
       "      <th>SL_Candidate</th>\n",
       "      <th>#InactiveSamples</th>\n",
       "      <th>#Samples</th>\n",
       "      <th>PValue</th>\n",
       "      <th>FDR</th>\n",
       "      <th>Tissue</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Inactive</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [Inactive, InactiveDB, SL_Candidate, #InactiveSamples, #Samples, PValue, FDR, Tissue]\n",
       "Index: []"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sof_CCLE = SurvivalOfFittest(client, 'SL', \"CCLE\", gene_list,  percentile_threshold, cn_threshold, pval_correction, fdr_level, list(tumor_type.value), input_mutations)\n",
    "try: \n",
    "    sof_ccle_intermediate_report=sof_CCLE.loc[(sof_CCLE['FDR'] < p_threshold),]\n",
    "    sof_ccle_report=sof_ccle_intermediate_report.groupby('Inactive').apply(lambda x: x.sort_values('FDR'))\n",
    "except:\n",
    "    sof_ccle_report=pd.DataFrame()\n",
    "    print(\"No results returned.\")\n",
    "sof_ccle_report"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The same analysis/output  with Pancancer Atlas Survival of Fittest analysis except for the dataset used (CCLE). "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 4.3. Functional examination inference module\n",
    "\n",
    "The rationale for the functional examination inference module is that if the synthetic lethal partner of a gene is inactive in a given sample, subsequent inactivation of that gene will be lethal. Therefore, for a gene of interest, we first defined two groups for the test, one in which the gene was inactive and the other in which it was not. We then performed a one-sided Wilcoxon rank-sum (Mann-Whitney U) test on the knockdown/knockout sensitivity of candidate synthetic lethal pairs of interest. Lower viability that is associated with higher knockout/knockdown sensitivity is an indicator of a potential SLI. The synthetic lethal pairs for  whom the test result P value was lower than 0.05 were returned. This inference procedure was applied to the gene-dependency scores or gene effect scores for the shRNA and CRISPR datasets separately. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "4.3.1. CRISPR based functional examination inference module"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>Inactive</th>\n",
       "      <th>InactiveDB</th>\n",
       "      <th>SL_Candidate</th>\n",
       "      <th>#InactiveSamples</th>\n",
       "      <th>#Samples</th>\n",
       "      <th>PValue</th>\n",
       "      <th>FDR</th>\n",
       "      <th>Tissue</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Inactive</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">ARID1A</th>\n",
       "      <th>0</th>\n",
       "      <td>ARID1A</td>\n",
       "      <td>ARID1A</td>\n",
       "      <td>ARID1B</td>\n",
       "      <td>100</td>\n",
       "      <td>771</td>\n",
       "      <td>1.114187e-09</td>\n",
       "      <td>0.000020</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ARID1A</td>\n",
       "      <td>ARID1A</td>\n",
       "      <td>CCDC102A</td>\n",
       "      <td>101</td>\n",
       "      <td>782</td>\n",
       "      <td>6.694568e-07</td>\n",
       "      <td>0.012130</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ARID1A</td>\n",
       "      <td>ARID1A</td>\n",
       "      <td>ARHGEF7</td>\n",
       "      <td>101</td>\n",
       "      <td>782</td>\n",
       "      <td>1.747027e-06</td>\n",
       "      <td>0.031654</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ARID1A</td>\n",
       "      <td>ARID1A</td>\n",
       "      <td>RPL22L1</td>\n",
       "      <td>101</td>\n",
       "      <td>782</td>\n",
       "      <td>2.011955e-06</td>\n",
       "      <td>0.036455</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ARID1A</td>\n",
       "      <td>ARID1A</td>\n",
       "      <td>PAK2</td>\n",
       "      <td>101</td>\n",
       "      <td>782</td>\n",
       "      <td>4.170099e-06</td>\n",
       "      <td>0.075558</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">BRCA2</th>\n",
       "      <th>3873</th>\n",
       "      <td>BRCA2</td>\n",
       "      <td>BRCA2</td>\n",
       "      <td>IGLON5</td>\n",
       "      <td>71</td>\n",
       "      <td>782</td>\n",
       "      <td>4.970267e-02</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3880</th>\n",
       "      <td>BRCA2</td>\n",
       "      <td>BRCA2</td>\n",
       "      <td>FRMD4A</td>\n",
       "      <td>71</td>\n",
       "      <td>782</td>\n",
       "      <td>4.981589e-02</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3883</th>\n",
       "      <td>BRCA2</td>\n",
       "      <td>BRCA2</td>\n",
       "      <td>DSTYK</td>\n",
       "      <td>71</td>\n",
       "      <td>782</td>\n",
       "      <td>4.987257e-02</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3886</th>\n",
       "      <td>BRCA2</td>\n",
       "      <td>BRCA2</td>\n",
       "      <td>CCDC170</td>\n",
       "      <td>71</td>\n",
       "      <td>782</td>\n",
       "      <td>4.992931e-02</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3887</th>\n",
       "      <td>BRCA2</td>\n",
       "      <td>BRCA2</td>\n",
       "      <td>TFF1</td>\n",
       "      <td>71</td>\n",
       "      <td>782</td>\n",
       "      <td>4.998610e-02</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3888 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              Inactive InactiveDB SL_Candidate  #InactiveSamples  #Samples  \\\n",
       "Inactive                                                                     \n",
       "ARID1A   0      ARID1A     ARID1A       ARID1B               100       771   \n",
       "         1      ARID1A     ARID1A     CCDC102A               101       782   \n",
       "         2      ARID1A     ARID1A      ARHGEF7               101       782   \n",
       "         3      ARID1A     ARID1A      RPL22L1               101       782   \n",
       "         5      ARID1A     ARID1A         PAK2               101       782   \n",
       "...                ...        ...          ...               ...       ...   \n",
       "BRCA2    3873    BRCA2      BRCA2       IGLON5                71       782   \n",
       "         3880    BRCA2      BRCA2       FRMD4A                71       782   \n",
       "         3883    BRCA2      BRCA2        DSTYK                71       782   \n",
       "         3886    BRCA2      BRCA2      CCDC170                71       782   \n",
       "         3887    BRCA2      BRCA2         TFF1                71       782   \n",
       "\n",
       "                     PValue       FDR         Tissue  \n",
       "Inactive                                              \n",
       "ARID1A   0     1.114187e-09  0.000020  ['pancancer']  \n",
       "         1     6.694568e-07  0.012130  ['pancancer']  \n",
       "         2     1.747027e-06  0.031654  ['pancancer']  \n",
       "         3     2.011955e-06  0.036455  ['pancancer']  \n",
       "         5     4.170099e-06  0.075558  ['pancancer']  \n",
       "...                     ...       ...            ...  \n",
       "BRCA2    3873  4.970267e-02  1.000000  ['pancancer']  \n",
       "         3880  4.981589e-02  1.000000  ['pancancer']  \n",
       "         3883  4.987257e-02  1.000000  ['pancancer']  \n",
       "         3886  4.992931e-02  1.000000  ['pancancer']  \n",
       "         3887  4.998610e-02  1.000000  ['pancancer']  \n",
       "\n",
       "[3888 rows x 8 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "crispr_result = FunctionalExamination(client,'SL', \"CRISPR\", gene_list, percentile_threshold, \n",
    "                                      cn_threshold, pval_correction,  fdr_level, list(tumor_type.value), input_mutations )\n",
    "try:\n",
    "    crispr_intermediate_report=crispr_result.loc[(crispr_result['PValue'] < p_threshold),]\n",
    "    crispr_report=crispr_intermediate_report.groupby('Inactive').apply(lambda x: x.sort_values('PValue'))\n",
    "except:\n",
    "    crispr_report=pd.DataFrame()\n",
    "    print(\"No results returned.\")\n",
    "crispr_report   "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Inactive/Overactive** (for SL and for SDL respectively): The gene name coming from the input list.\n",
    "\n",
    "**InactiveDB/OveractiveDB**: The corresponding gene name in the dataset (Pancancer Atlas).\n",
    "\n",
    "**SL_Candidate**: The Candidate SL pair of the gene in the same row. \n",
    "\n",
    "**#InactiveSamples:** The number of inactive samples.\n",
    "\n",
    "**#Samples:** The total number of samples that the analysis is performed on. \n",
    "\n",
    "**PValue:** The p values for the Wilcoxon umpaired one sided test on CRISPR gene essentiality data.\n",
    "\n",
    "**FDR:** Corrected p value\n",
    "\n",
    "**Tissue:** The tissue(s) that the analysis has been performed."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<br>\n",
    "4.3.2. shRNA based functional examination inference module"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>Inactive</th>\n",
       "      <th>InactiveDB</th>\n",
       "      <th>SL_Candidate</th>\n",
       "      <th>#InactiveSamples</th>\n",
       "      <th>#Samples</th>\n",
       "      <th>PValue</th>\n",
       "      <th>FDR</th>\n",
       "      <th>Tissue</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Inactive</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">ARID1A</th>\n",
       "      <th>0</th>\n",
       "      <td>ARID1A</td>\n",
       "      <td>ARID1A</td>\n",
       "      <td>SIRPA</td>\n",
       "      <td>109</td>\n",
       "      <td>651</td>\n",
       "      <td>0.000006</td>\n",
       "      <td>0.108836</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ARID1A</td>\n",
       "      <td>ARID1A</td>\n",
       "      <td>BTF3</td>\n",
       "      <td>110</td>\n",
       "      <td>651</td>\n",
       "      <td>0.000008</td>\n",
       "      <td>0.145394</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ARID1A</td>\n",
       "      <td>ARID1A</td>\n",
       "      <td>KCNK1</td>\n",
       "      <td>109</td>\n",
       "      <td>649</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.169046</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ARID1A</td>\n",
       "      <td>ARID1A</td>\n",
       "      <td>VAV2</td>\n",
       "      <td>110</td>\n",
       "      <td>651</td>\n",
       "      <td>0.000022</td>\n",
       "      <td>0.380350</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ARID1A</td>\n",
       "      <td>ARID1A</td>\n",
       "      <td>RPS6KC1</td>\n",
       "      <td>109</td>\n",
       "      <td>649</td>\n",
       "      <td>0.000031</td>\n",
       "      <td>0.527259</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">BRCA2</th>\n",
       "      <th>3227</th>\n",
       "      <td>BRCA2</td>\n",
       "      <td>BRCA2</td>\n",
       "      <td>PPP1R2P3</td>\n",
       "      <td>57</td>\n",
       "      <td>506</td>\n",
       "      <td>0.049892</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3228</th>\n",
       "      <td>BRCA2</td>\n",
       "      <td>BRCA2</td>\n",
       "      <td>SLC43A3</td>\n",
       "      <td>57</td>\n",
       "      <td>506</td>\n",
       "      <td>0.049892</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3229</th>\n",
       "      <td>BRCA2</td>\n",
       "      <td>BRCA2</td>\n",
       "      <td>SGO2</td>\n",
       "      <td>57</td>\n",
       "      <td>506</td>\n",
       "      <td>0.049892</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3230</th>\n",
       "      <td>BRCA2</td>\n",
       "      <td>BRCA2</td>\n",
       "      <td>HAND2</td>\n",
       "      <td>76</td>\n",
       "      <td>651</td>\n",
       "      <td>0.049903</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3234</th>\n",
       "      <td>BRCA2</td>\n",
       "      <td>BRCA2</td>\n",
       "      <td>ACBD5</td>\n",
       "      <td>76</td>\n",
       "      <td>651</td>\n",
       "      <td>0.049970</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3236 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              Inactive InactiveDB SL_Candidate  #InactiveSamples  #Samples  \\\n",
       "Inactive                                                                     \n",
       "ARID1A   0      ARID1A     ARID1A        SIRPA               109       651   \n",
       "         1      ARID1A     ARID1A         BTF3               110       651   \n",
       "         3      ARID1A     ARID1A        KCNK1               109       649   \n",
       "         4      ARID1A     ARID1A         VAV2               110       651   \n",
       "         5      ARID1A     ARID1A      RPS6KC1               109       649   \n",
       "...                ...        ...          ...               ...       ...   \n",
       "BRCA2    3227    BRCA2      BRCA2     PPP1R2P3                57       506   \n",
       "         3228    BRCA2      BRCA2      SLC43A3                57       506   \n",
       "         3229    BRCA2      BRCA2         SGO2                57       506   \n",
       "         3230    BRCA2      BRCA2        HAND2                76       651   \n",
       "         3234    BRCA2      BRCA2        ACBD5                76       651   \n",
       "\n",
       "                 PValue       FDR         Tissue  \n",
       "Inactive                                          \n",
       "ARID1A   0     0.000006  0.108836  ['pancancer']  \n",
       "         1     0.000008  0.145394  ['pancancer']  \n",
       "         3     0.000010  0.169046  ['pancancer']  \n",
       "         4     0.000022  0.380350  ['pancancer']  \n",
       "         5     0.000031  0.527259  ['pancancer']  \n",
       "...                 ...       ...            ...  \n",
       "BRCA2    3227  0.049892  1.000000  ['pancancer']  \n",
       "         3228  0.049892  1.000000  ['pancancer']  \n",
       "         3229  0.049892  1.000000  ['pancancer']  \n",
       "         3230  0.049903  1.000000  ['pancancer']  \n",
       "         3234  0.049970  1.000000  ['pancancer']  \n",
       "\n",
       "[3236 rows x 8 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "shRNA_result = FunctionalExamination(client, 'SL', \"shRNA\", gene_list , percentile_threshold, \\\n",
    "                                     cn_threshold, pval_correction,  fdr_level, list(tumor_type.value),input_mutations)\n",
    "try:\n",
    "    shRNA_intermediate_report=shRNA_result.loc[(shRNA_result['PValue'] < p_threshold),]\n",
    "    shRNA_report=shRNA_intermediate_report.groupby('Inactive').apply(lambda x: x.sort_values('PValue'))\n",
    "    \n",
    "except:\n",
    "    shRNA_report=pd.DataFrame()\n",
    "    print(\"No results returned.\")\n",
    "shRNA_report"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The same analysis/output with CRISPR Functional Examination analysis except for the dataset used (shRNA- gene dependency). "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 5. Integration of results"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "5.1. Integration of the pairwise Co-expression gene co-expression results on Pancancer and CCLE\n",
    "\n",
    "The union of results from PanCancer Atlas and CCLE was used. \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Inactive</th>\n",
       "      <th>SL_Candidate</th>\n",
       "      <th>FDR0</th>\n",
       "      <th>FDR1</th>\n",
       "      <th>Tissue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ARID1A</td>\n",
       "      <td>SPEN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.800289e-246</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1477</th>\n",
       "      <td>ARID1A</td>\n",
       "      <td>MCM4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6.419231e-112</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1476</th>\n",
       "      <td>ARID1A</td>\n",
       "      <td>ZNF346</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.861848e-112</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1475</th>\n",
       "      <td>ARID1A</td>\n",
       "      <td>AKAP8L</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.725635e-112</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1474</th>\n",
       "      <td>ARID1A</td>\n",
       "      <td>MDM4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.436705e-112</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>779</th>\n",
       "      <td>BRCA2</td>\n",
       "      <td>FAM102B</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>780</th>\n",
       "      <td>BRCA2</td>\n",
       "      <td>RPAP2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>781</th>\n",
       "      <td>BRCA2</td>\n",
       "      <td>STAG1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>765</th>\n",
       "      <td>BRCA2</td>\n",
       "      <td>C9orf41</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2440</th>\n",
       "      <td>BRCA2</td>\n",
       "      <td>CEP295</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.988797e-79</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2441 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Inactive SL_Candidate  FDR0           FDR1         Tissue\n",
       "0      ARID1A         SPEN   0.0  7.800289e-246  ['pancancer']\n",
       "1477   ARID1A         MCM4   NaN  6.419231e-112  ['pancancer']\n",
       "1476   ARID1A       ZNF346   NaN  4.861848e-112  ['pancancer']\n",
       "1475   ARID1A       AKAP8L   NaN  3.725635e-112  ['pancancer']\n",
       "1474   ARID1A         MDM4   NaN  3.436705e-112  ['pancancer']\n",
       "...       ...          ...   ...            ...            ...\n",
       "779     BRCA2      FAM102B   0.0            NaN  ['pancancer']\n",
       "780     BRCA2        RPAP2   0.0            NaN  ['pancancer']\n",
       "781     BRCA2        STAG1   0.0            NaN  ['pancancer']\n",
       "765     BRCA2      C9orf41   0.0            NaN  ['pancancer']\n",
       "2440    BRCA2       CEP295   NaN   3.988797e-79  ['pancancer']\n",
       "\n",
       "[2441 rows x 5 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "try:\n",
    "    coexpression_result = UnionResults([coexp_pancancer_report, coexp_CCLE_report],'SL', ['FDR', 'FDR'],  list(tumor_type.value))\n",
    "    coexpression_result=coexpression_result.sort_values('Inactive')\n",
    "except:\n",
    "    coexpression_result=pd.DataFrame()\n",
    "    print(\"No Result From Pairwise Co-expression Inference Procedure\")\n",
    "    \n",
    "coexpression_result"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Inactive:** The Inactive gene name\n",
    "\n",
    "**SL_Candidate/SDL_Candidate:** The candidate SL or SDL partner \n",
    "\n",
    "**p values or FDRs** The aggregated p value or FDR with Fishers method\n",
    "\n",
    "**Tissue:** The tissue(s) that the analysis has been performed.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<br>\n",
    "5.2. Integration of Survival of Fittest results on Pancancer and CCLE\n",
    "\n",
    "The union of results from PanCancer Atlas and CCLE was used. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Inactive</th>\n",
       "      <th>SL_Candidate</th>\n",
       "      <th>FDR0</th>\n",
       "      <th>Tissue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ARID1A</td>\n",
       "      <td>ENTPD7</td>\n",
       "      <td>5.161072e-10</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>649</th>\n",
       "      <td>ARID1A</td>\n",
       "      <td>PTPN20A</td>\n",
       "      <td>6.916944e-04</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>650</th>\n",
       "      <td>ARID1A</td>\n",
       "      <td>PTPN20B</td>\n",
       "      <td>6.916944e-04</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>651</th>\n",
       "      <td>ARID1A</td>\n",
       "      <td>IPMK</td>\n",
       "      <td>7.616928e-04</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>652</th>\n",
       "      <td>ARID1A</td>\n",
       "      <td>TFAM</td>\n",
       "      <td>7.727423e-04</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1590</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>ACAN</td>\n",
       "      <td>1.153027e-04</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1589</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>KIF23</td>\n",
       "      <td>1.148171e-04</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1588</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>MFGE8</td>\n",
       "      <td>1.141844e-04</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1586</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>AEN</td>\n",
       "      <td>1.085066e-04</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2820</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>LENG8</td>\n",
       "      <td>4.994935e-02</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2821 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Inactive SL_Candidate          FDR0         Tissue\n",
       "0      ARID1A       ENTPD7  5.161072e-10  ['pancancer']\n",
       "649    ARID1A      PTPN20A  6.916944e-04  ['pancancer']\n",
       "650    ARID1A      PTPN20B  6.916944e-04  ['pancancer']\n",
       "651    ARID1A         IPMK  7.616928e-04  ['pancancer']\n",
       "652    ARID1A         TFAM  7.727423e-04  ['pancancer']\n",
       "...       ...          ...           ...            ...\n",
       "1590    BRCA1         ACAN  1.153027e-04  ['pancancer']\n",
       "1589    BRCA1        KIF23  1.148171e-04  ['pancancer']\n",
       "1588    BRCA1        MFGE8  1.141844e-04  ['pancancer']\n",
       "1586    BRCA1          AEN  1.085066e-04  ['pancancer']\n",
       "2820    BRCA1        LENG8  4.994935e-02  ['pancancer']\n",
       "\n",
       "[2821 rows x 4 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "try:\n",
    "    sof_result = UnionResults([sof_ccle_report, sof_pancancer_report],  'SL', ['FDR', 'FDR'], list(tumor_type.value))\n",
    "    sof_result=sof_result.sort_values('Inactive')\n",
    "except:\n",
    "    sof_result=pd.DataFrame()\n",
    "    print(\"No Result From Survival of Fittest Inference Procedure\")\n",
    "sof_result    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Inactive:** The Inactive gene name\n",
    "\n",
    "**SL_Candidate/SDL_Candidate:** The candidate SL or SDL partner \n",
    "\n",
    "**p values or FDRs** The aggregated p value or FDR with Fishers method\n",
    "\n",
    "**Tissue:** The tissue(s) that the analysis has been performed.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<br>\n",
    "5.3. Integration of shRNA and CRISPR based functional examination inference module.\n",
    "\n",
    "We reported the union of results from shRNA and CRISPR-based datasets. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Inactive</th>\n",
       "      <th>SL_Candidate</th>\n",
       "      <th>PValue0</th>\n",
       "      <th>PValue1</th>\n",
       "      <th>Tissue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ARID1A</td>\n",
       "      <td>ARID1B</td>\n",
       "      <td>1.114187e-09</td>\n",
       "      <td>0.000350</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4850</th>\n",
       "      <td>ARID1A</td>\n",
       "      <td>GALNT16</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.029362</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4849</th>\n",
       "      <td>ARID1A</td>\n",
       "      <td>KAT14</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.029348</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4848</th>\n",
       "      <td>ARID1A</td>\n",
       "      <td>TTC23L</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.029338</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4847</th>\n",
       "      <td>ARID1A</td>\n",
       "      <td>TMEM259</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.029251</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3426</th>\n",
       "      <td>BRCA2</td>\n",
       "      <td>C17orf78</td>\n",
       "      <td>3.083288e-02</td>\n",
       "      <td>NaN</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3427</th>\n",
       "      <td>BRCA2</td>\n",
       "      <td>C9orf3</td>\n",
       "      <td>3.087125e-02</td>\n",
       "      <td>NaN</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3428</th>\n",
       "      <td>BRCA2</td>\n",
       "      <td>PFDN4</td>\n",
       "      <td>3.087125e-02</td>\n",
       "      <td>0.031764</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3420</th>\n",
       "      <td>BRCA2</td>\n",
       "      <td>DYNC1I2</td>\n",
       "      <td>3.064162e-02</td>\n",
       "      <td>0.032130</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6847</th>\n",
       "      <td>BRCA2</td>\n",
       "      <td>ACBD5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.049970</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6848 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Inactive SL_Candidate       PValue0   PValue1         Tissue\n",
       "0      ARID1A       ARID1B  1.114187e-09  0.000350  ['pancancer']\n",
       "4850   ARID1A      GALNT16           NaN  0.029362  ['pancancer']\n",
       "4849   ARID1A        KAT14           NaN  0.029348  ['pancancer']\n",
       "4848   ARID1A       TTC23L           NaN  0.029338  ['pancancer']\n",
       "4847   ARID1A      TMEM259           NaN  0.029251  ['pancancer']\n",
       "...       ...          ...           ...       ...            ...\n",
       "3426    BRCA2     C17orf78  3.083288e-02       NaN  ['pancancer']\n",
       "3427    BRCA2       C9orf3  3.087125e-02       NaN  ['pancancer']\n",
       "3428    BRCA2        PFDN4  3.087125e-02  0.031764  ['pancancer']\n",
       "3420    BRCA2      DYNC1I2  3.064162e-02  0.032130  ['pancancer']\n",
       "6847    BRCA2        ACBD5           NaN  0.049970  ['pancancer']\n",
       "\n",
       "[6848 rows x 5 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "try:\n",
    "    functional_screening_result = UnionResults([crispr_report, shRNA_report],'SL', ['PValue', 'PValue'], list(tumor_type.value))\n",
    "    functional_screening_result=functional_screening_result.sort_values('Inactive')\n",
    "    \n",
    "except:\n",
    "    functional_screening_result=pd.DataFrame()\n",
    "    print(\"No Result From Functional Examination Inference Procedure\")\n",
    "functional_screening_result    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Inactive:** The Inactive gene name\n",
    "\n",
    "**SL_Candidate/SDL_Candidate:** The candidate SL or SDL partner \n",
    "\n",
    "**p values or FDRs** The aggregated p value or FDR with Fishers method\n",
    "\n",
    "**Tissue:** The tissue(s) that the analysis has been performed.\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<br>\n",
    "5.4. Merging the results from all three inference procedures\n",
    "\n",
    "The intersection of SL pairs from different inference procedures compose the final list. \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Inactive</th>\n",
       "      <th>SL_Candidate</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ARID1A</td>\n",
       "      <td>SUZ12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ARID1A</td>\n",
       "      <td>LCOR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ARID1A</td>\n",
       "      <td>PHF12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>HNRNPM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>ANP32A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Inactive SL_Candidate\n",
       "0   ARID1A        SUZ12\n",
       "1   ARID1A         LCOR\n",
       "2   ARID1A        PHF12\n",
       "3    BRCA1       HNRNPM\n",
       "4    BRCA1       ANP32A"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "try:\n",
    "    all_merged_results = MergeResults([coexpression_result, sof_result, functional_screening_result], 'SL',  list(tumor_type.value))\n",
    "    all_merged_results=all_merged_results.sort_values('Inactive')\n",
    "except:\n",
    "    all_merged_results=pd.DataFrame()\n",
    "    print(\"No results found\")\n",
    "all_merged_results"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Inactive/Overactive**: The gene from the user input\n",
    "\n",
    "\n",
    "**SL/SDL candidate**: The inferred SL/SDL partner of the gene in the same row. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Results can also be saved into the excel files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "WriteToExcel(\"DAISY_SL_results.xlsx\", [all_merged_results], [\"final results\"])"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
